crates/bevy_ecs/src/query/state.rs

```diff
@@ -338,12 +348,20 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
             let archetype_index = archetype.id().index();
             if !self.matched_archetypes.contains(archetype_index) {
                 self.matched_archetypes.grow_and_insert(archetype_index);
-                self.matched_archetype_ids.push(archetype.id());
+                if !D::IS_DENSE || !F::IS_DENSE {
+                    self.matched_storage_ids.push(StorageId {
+                        archetype_id: archetype.id(),
+                    });
+                }
             }
             let table_index = archetype.table_id().as_usize();
             if !self.matched_tables.contains(table_index) {
                 self.matched_tables.grow_and_insert(table_index);
-                self.matched_table_ids.push(archetype.table_id());
+                if D::IS_DENSE && F::IS_DENSE {
```

Comment:
Do we still also need separate `self.matched_tables` and `self.matched_archetypes` bitsets?
It looks like only one of the two will be used, depending on the value of `D::IS_DENSE` and `F::IS_DENSE`

Maybe it's still needed for things like `join`?

Reply:
Yep exactly, join needs them, and get_unchecked_manual needs specifically matched_archetypes (right now).